org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/model/AbstractTMModel.java

```diff
@@ -128,170 +70,203 @@ public int getSize() {
 		 * @param name
 		 *            the thread's name
 		 */
-		public BackgroundThread(String name) {
+		public TokenizerThread(String name, AbstractTMModel model) {
 			super(name);
+			this.model = model;
+			this.tokenizer = new Tokenizer(model.getGrammar());
 			setPriority(Thread.MIN_PRIORITY);
 			setDaemon(true);
 		}
 
-		/**
-		 * Suspends the caller of this method until this background thread has
-		 * emptied the dirty region queue.
-		 */
-		public void suspendCallerWhileDirty() {
-			boolean isDirty;
-			do {
-				synchronized (fDirtyRegionQueue) {
-					isDirty = fDirtyRegionQueue.getSize() > 0;
-					if (isDirty) {
+		@Override
+		public void run() {
+			if (interrupted())
+				return;
+
+			// initialize
+			int linesLength = model.getNumberOfLines();
+			for (int line = 0; line < linesLength; line++) {
+				try {
+					model.lines.addLine(line);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+			}
+			model.lines.forEach(ModelLine::resetTokenizationState);
+			model.lines.get(0).setState(tokenizer.getInitialState());
+			model._invalidLineStartIndex = 0;
+			
+			while (!interrupted()) {
+				if (model.lines.getSize() == 0) {
+					synchronized (model.lines) {
 						try {
-							fDirtyRegionQueue.wait();
+							model.lines.wait(500); // release synchronized lock at least every 500ms
 						} catch (InterruptedException x) {
 						}
 					}
 				}
-			} while (isDirty);
-		}
-
-		/**
-		 * Reset the background thread as the text viewer has been changed,
-		 */
-		public void reset() {
+				while (model.lines.getSize() >= 0 && model._invalidLineStartIndex < model.lines.getSize()) {

```

Comment:
Here the code which works for me:

```
boolean init = true;
while (!interrupted()) {
	if (!init) {
		synchronized (model.lines) {
			try {
				model.lines.wait(500); // release synchronized lock at least every 500ms
			} catch (InterruptedException x) {
				Thread.currentThread().interrupt();
			}
		}
	}
	init = false;
	while (model.lines.getSize() >= 0 && model._invalidLineStartIndex < model.lines.getSize()) {
		if (interrupted()) {
			return;
		}
		this._revalidateTokensNow(null);
	}
}
```

wait is done every time except when editor is opened.

The call of Thread.currentThread().interrupt(); is very important otherwise the thread is not really interrupted. (It was my problem I think with thread was not interrupted)